lib/babe/babe.go

```diff
@@ -497,6 +497,21 @@ func (b *Service) handleSlot(epoch, slotNum uint64,
 		return err
 	}
 
+	slotNumOfTheBestBlock, err := b.blockState.GetSlotForBlock(parentHeader.Hash())
+	if err != nil {
+		return fmt.Errorf("could not get slot for block %s: %w", parentHeader.Hash(), err)
+	}
+
+	if slotNumOfTheBestBlock == slotNum {
+		// pick parent of best block instead to handle slot
+		newParentHeader, err := b.blockState.GetHeader(parentHeader.ParentHash)
+		if err != nil {
+			return fmt.Errorf("could not get header for hash %s: %w", parentHeader.ParentHash, err)
```

Comment:
nit:

- the hash should be added as context by `GetHeader`, not here
- removing the `could not` prefix
- adding the `grandparent` context? I guess parent hash of parent header is grandparent right?

```suggestion
			return fmt.Errorf("getting grandparent header: %w", err)
```

Reply:
> since it builds up a repetition for deeply wrapped errors

I don't think so, would be much easier to read eg. `cannot build block: cannot get slot number: failed to get block from the tree` than `cannot build block: getting slot number: retrieving block from the tree`, I think writing better log messages is needed but in this case I think we should keep as it is

Reply:
I don't see how `cannot build block: cannot get slot number: (more cannots): retrieving block from the tree` is not worst than `building block: getting slot number: (more verbings): retrieving block from the tree failed`. Only the wrapped (bottom) error should state is failed I would say.

I was a die hard fan or prefixing all wrappings with `cannot` for months if not years to be honest, but seeing horribly long repetitive cannot wrappings in error logs, I decided it was time for change😿 

Let's see what others feel like, I'll send the link to this on Slack, such that we use the convention for errors and error wrappings.

Reply:
> adding the grandparent context? I guess parent hash of parent header is grandparent right?

can't use the word grandparent. Basically we are looking for a block on which we can put our new block (parent of the block to be minted). So it is always parent.

Reply:
> the hash should be added as context by GetHeader, not here

blockState.GetHeader doesn't do all that. I have passed hash in the argument, it is not unreasonable to add that here in the error.

> Let's see what others feel like, I'll send the link to this on Slack, such that we use the convention for errors and error wrappings.

I am happy with the current style of error wording. don't have a problem with `cannot` or `could not`.

Reply:
> I am happy with the current style of error wording. don't have a problem with cannot or could not.

Sure up to you. On the other hand, really the hash should not be added here, or it will be twice in the error message. Only the last function handling the hash should wrap it as context to the error. Especially important since hashes are not that short.

Reply:
> really the hash should not be added here, or it will be twice in the error message.

But none of the errors in GetHeader are wrapping hash in them 

https://github.com/ChainSafe/gossamer/blob/b56d7f91fd4f9603473f834b94d5e26b34ebd306/dot/state/block.go#L202-L233